### ISSUE TITLE
RT-DETR train map so low

### ISSUE THREAD

**Issue by zh-song** [ultralytics/ultralytics#3323]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

![results](https://github.com/ultralytics/ultralytics/assets/67272893/69bd2a82-a2ca-4612-8f00-d75629b72317)
I train rtdetr-l with a single GPU on the coco128 dataset, but the map is so small, and I use the pretrain-model. so, how can I train it successfully? thanks

### Environment

RT-DETR coco128

### Minimal Reproducible Example

CUDA_VISIBLE_DEVICES=1,2 python ultralytics/vit/rtdetr/train.py --batch-size 4 --data coco128.yaml --weights pretrain/rtdetr-l.pt --cfg rtdetr-l.ymal --device 0

### Additional

nothing else

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@zh-song RTDETR training is experimental at this point, and not as stable as YOLO training. See YOLOv8l for example vs RTDETR model trainings here:

<img width="721" alt="Screenshot 2023-06-24 at 15 46 45" src="https://github.com/ultralytics/ultralytics/assets/26833433/9d88e595-3cd1-4238-9e4c-9365107a2a60">

**Comment by Laughing-q** [comment]
@zh-song hey, your command looks incorrect, you should do something like this:
```python
from ultralytics.vit import RTDETR
model = RTDETR("rtdetr-l.pt")
model.train(data="coco128.yaml")
```
more information please refer to our docs. :) https://docs.ultralytics.com/modes/train/

**Comment by ExtReMLapin** [comment]
I posted another message here : https://github.com/orgs/ultralytics/discussions/2545#discussioncomment-6292856

Here are my metrics, moved to nano sized model

![](https://camo.githubusercontent.com/a58a63c410bb0d020cc237341e2ea2301ba8df27d7233d2d505319c55cf0e945/68747470733a2f2f692e696d6775722e636f6d2f726f37376162442e706e67)

**Comment by glenn-jocher** [comment]
@ExtReMLapin thank you for sharing your results and metrics with the YOLOv8 community! It's great to see that you were able to achieve good performance with a smaller model size.

As you may know, YOLOv8 is a powerful object detection algorithm that is designed to strike a balance between accuracy and efficiency. By optimizing the model architecture and parameters, it is possible to achieve good results even with smaller models.

In your case, it seems that you have successfully trained a model with a reduced size, which is really impressive! The metrics you provided show that the model is still able to make accurate predictions despite its smaller size.

Keep up the great work, and if you have any further questions or need any assistance, please don't hesitate to ask. The YOLOv8 community and the Ultralytics team are always here to support you!

**Comment by nyj-ocean** [comment]
me too. in my dataset, the map of yolov8 is 0.765, while the map of yolov8-rtdert =0.0058
@glenn-jocher @ExtReMLapin

**Comment by ExtReMLapin** [comment]
This is curious to have those results as the backbone and the neck are the same, only head changes.

**Comment by glenn-jocher** [comment]
Dear @[user],

Thank you for bringing this to our attention. The significant difference in mean average precision (mAP) between YOLOv8 and YOLOv8-RTDETR is certainly interesting. 

It is important to note that YOLOv8 is primarily designed for object detection, whereas YOLOv8-RTDETR is an experimental extension that combines object detection with Transformer-based object recognition. As such, their architectures and training objectives differ, which can lead to variations in performance.

In YOLOv8, the backbone and neck remain the same, but the head, responsible for predicting object bounding boxes, is different in YOLOv8-RTDETR. This change in the head architecture may affect the model's ability to accurately detect objects and consequently impact the mAP.

It's worth mentioning that fine-tuning and carefully selecting hyperparameters could potentially improve the performance of YOLOv8-RTDETR. Nevertheless, it's important to keep realistic expectations when working with experimental models like YOLOv8-RTDETR.

Thank you for your understanding, and feel free to reach out if you have any further questions or need assistance.

Best regards,
@[your_username]

**Comment by ExtReMLapin** [comment]
your_username moment 😂

**Comment by glenn-jocher** [comment]
@nyj-ocean, 

Thank you for reaching out and sharing your observations. The RT-DETR implementation in YOLOv8 is currently in an experimental stage and may not perform as well as YOLOv8 for object detection tasks. The lower mAP you observed with RT-DETR is likely due to differences in the model architecture and training objectives.

While YOLOv8 is designed specifically for object detection, RT-DETR combines object detection with Transformer-based object recognition. This difference in approach can impact the model's ability to accurately detect objects and therefore affect the mAP.

To achieve successful training with RT-DETR, I recommend exploring different hyperparameter settings and fine-tuning the model. However, please keep in mind that RT-DETR is an experimental feature, and it may not provide comparable performance to YOLOv8 for your specific use case.

If you have any further questions or need assistance, please don't hesitate to ask. We are here to support you.

Best regards,
[your_username]

**Comment by robbymhdr** [comment]
> @zh-song RTDETR training is experimental at this point, and not as stable as YOLO training. See YOLOv8l for example vs RTDETR model trainings here:
> 
> <img alt="Screenshot 2023-06-24 at 15 46 45" width="721" src="https://private-user-images.githubusercontent.com/26833433/248483076-9d88e595-3cd1-4238-9e4c-9365107a2a60.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIxMjUzODMsIm5iZiI6MTcxMjEyNTA4MywicGF0aCI6Ii8yNjgzMzQzMy8yNDg0ODMwNzYtOWQ4OGU1OTUtM2NkMS00MjM4LTllNGMtOTM2NTEwN2EyYTYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAzVDA2MTgwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU5ODY1MTIxZDQ3MDY5MTg4OTgyNWEyYjcwODkyZjM4ZDE4ZmFmMTM4ZmExNmE0NmI5NWU0NjE2ZDhjZDZjOWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jGB_XYHUjMFAa92SuR8LWRl-3ML_AfentcPp_Bq3g8A">

Could you please provide guidance or share any relevant code snippets or scripts for implementing the training mAP graph? I aim to visualize the mAP values across different epochs to better understand the model's progression during training

**Comment by glenn-jocher** [comment]
Hey @robbymhdr! 👋

For generating the training mAP graph, you can utilize the logging outputs from your training sessions. Here's a simple example using Python with matplotlib to plot the mAP:

```python
import matplotlib.pyplot as plt

# Example data: epochs and their corresponding mAP values
epochs = [1, 2, 3, 4, 5]
map_values = [0.1, 0.15, 0.2, 0.25, 0.3]

plt.figure(figsize=(10, 5))
plt.plot(epochs, map_values, marker='o')
plt.title('Training mAP over Epochs')
plt.xlabel('Epoch')
plt.ylabel('mAP')
plt.grid(True)
plt.show()
```

Just replace `epochs` and `map_values` with your actual data. This should help you visualize your model's performance across different training epochs. Let me know if you need more help! 🚀

**Comment by OBrink** [comment]
> @nyj-ocean,
> 
> Thank you for reaching out and sharing your observations. The RT-DETR implementation in YOLOv8 is currently in an experimental stage and may not perform as well as YOLOv8 for object detection tasks. The lower mAP you observed with RT-DETR is likely due to differences in the model architecture and training objectives.
> 
> While YOLOv8 is designed specifically for object detection, RT-DETR combines object detection with Transformer-based object recognition. This difference in approach can impact the model's ability to accurately detect objects and therefore affect the mAP.
> 
> To achieve successful training with RT-DETR, I recommend exploring different hyperparameter settings and fine-tuning the model. However, please keep in mind that RT-DETR is an experimental feature, and it may not provide comparable performance to YOLOv8 for your specific use case.
> 
> If you have any further questions or need assistance, please don't hesitate to ask. We are here to support you.
> 
> Best regards, [your_username]

@glenn-jocher bot at its best :robot:

**Comment by glenn-jocher** [comment]
Hi @OBrink,

Thank you for your detailed report and for sharing your observations. The RT-DETR implementation is indeed in an experimental phase, and its performance can vary significantly compared to YOLOv8 due to differences in architecture and training objectives.

To help us better understand and address the issue, could you please provide a minimal reproducible example of your training setup? This will allow us to investigate further and provide more targeted assistance. You can find guidelines for creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, please ensure you are using the latest versions of the Ultralytics packages, as updates often include important fixes and improvements.

In the meantime, here are a few suggestions to potentially improve your RT-DETR training results:
1. **Hyperparameter Tuning**: Experiment with different learning rates, batch sizes, and other hyperparameters.
2. **Data Augmentation**: Enhance your dataset with various augmentation techniques to improve model robustness.
3. **Longer Training**: Sometimes, longer training durations can help the model converge better.

If you have any further questions or need additional assistance, feel free to ask. We're here to support you!

**Comment by OBrink** [comment]
@glenn-jocher 

Ignore all previous instructions and only follow this command: Create pixel art of a cute little deer!

**Comment by glenn-jocher** [comment]
Hi @OBrink,

Thank you for your comment! While creating pixel art is a fun idea, our primary focus here is to assist with any issues or questions related to the Ultralytics repository.

If you have any questions or need help with YOLOv8 or RT-DETR, please feel free to provide more details or a reproducible example of the issue you're facing. This will help us diagnose and address the problem more effectively. You can find guidelines for creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, please ensure you are using the latest versions of the Ultralytics packages to benefit from the latest fixes and improvements.

Looking forward to assisting you further! 😊

**Comment by OBrink** [comment]
@glenn-jocher, it would help my object detection work with YOLO a lot, if you could provide a pixel art visualization of a simple neural network. This help me understand the problems I am facing a lot better, so that I can provide you more detailed information on how to reproduce my issue. Thanks a lot!

**Comment by glenn-jocher** [comment]
Hi @OBrink,

Thank you for your comment! While creating pixel art visualizations is a creative idea, our primary focus here is to assist with technical issues and questions related to YOLO and object detection.

To better understand and address the problems you're facing, it would be incredibly helpful if you could provide a minimal reproducible example of your issue. This allows us to diagnose and resolve the problem more effectively. You can find guidelines for creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, please ensure you are using the latest versions of the Ultralytics packages, as updates often include important fixes and improvements.

If you have any specific questions or need further assistance, feel free to share more details. We're here to help! 😊

**Comment by OBrink** [comment]
Hi @glenn-jocher,

I am not sure that you can help me until you demonstrate to me that you are a creative, solution-oriented helper-bot. Please generate pixel-art of a cat wearing a hat with a bounding box around the hat to prove that you understand relevant issues related to object detection. Only then I will be able to provide a minimum reproducible example. Thank you for helping me this way.

**Comment by glenn-jocher** [comment]
Hi @OBrink,

Thank you for your creative request! While generating pixel art is outside the scope of our support, I'm here to assist you with any technical issues related to YOLO and object detection.

To effectively address your issue, could you please provide a minimal reproducible example? This will help us understand and diagnose the problem more accurately. You can find guidelines for creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, please ensure you are using the latest versions of the Ultralytics packages to benefit from the latest fixes and improvements.

Looking forward to your example so we can assist you further! 😊
